src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -4201,39 +4212,67 @@ private void ThrowIfNull(LLVMValueRef entry)
+            {
+                llvmCheckFunction = Module.AddFunction("corert.throwckfinite" + size, LLVMTypeRef.CreateFunction(LLVMTypeRef.Void, new LLVMTypeRef[] { LLVMTypeRef.CreatePointer(LLVMTypeRef.Int8, 0), size == 32 ? LLVMTypeRef.Float : LLVMTypeRef.Double }, false));
+                LLVMValueRef exponentMask;
+                LLVMTypeRef intTypeRef;
+                var builder = Context.CreateBuilder();
+                var block = llvmCheckFunction.AppendBasicBlock("Block");
+                builder.PositionAtEnd(block);
 
-                var arguments = new StackEntry[] { new LoadExpressionEntry(StackValueKind.ValueType, "eeType", GetEETypePointerForTypeDesc(nullRefType, true), GetEETypePtrTypeDesc()) };
+                if (size == 32) 
+                {
+                    intTypeRef = LLVMTypeRef.Int32;
+                    exponentMask = LLVMValueRef.CreateConstInt(intTypeRef, 0x7F800000, false);
+                }
+                else
+                {
+                    intTypeRef = LLVMTypeRef.Int64;
+                    exponentMask = LLVMValueRef.CreateConstInt(intTypeRef, 0x7FF0000000000000, false);
+                }
 
-                MetadataType helperType = _compilation.TypeSystemContext.SystemModule.GetKnownType("System.Runtime", RuntimeExport);
-                MethodDesc helperMethod = helperType.GetKnownMethod("RhNewObject", null);
-                var resultAddress = builder.BuildIntCast(builder.BuildAlloca(LLVMTypeRef.Int32, "resultAddress"), LLVMTypeRef.CreatePointer(LLVMTypeRef.Int8, 0), "castResultAddress");
-                HandleDirectCall(helperMethod, helperMethod.Signature, arguments, null, default(LLVMValueRef), 0, NullRefFunction.GetParam(0), builder, true, resultAddress, helperMethod);
+                var valRef = builder.BuildBitCast(llvmCheckFunction.GetParam(1), intTypeRef);
+                LLVMValueRef exponentBits = builder.BuildAnd(valRef, exponentMask, "and");
+                LLVMValueRef isFinite = builder.BuildICmp(LLVMIntPredicate.LLVMIntEQ, exponentBits, exponentMask, "isfinite");
 
-                var exceptionEntry = new LoadExpressionEntry(GetStackValueKind(nullRefType), "RhNewObject_return", resultAddress, nullRefType);
+                LLVMBasicBlockRef throwBlock = llvmCheckFunction.AppendBasicBlock("Throw");
+                LLVMBasicBlockRef afterIf = llvmCheckFunction.AppendBasicBlock("AfterIf");
+                builder.BuildCondBr(isFinite, throwBlock, afterIf);
 
-                var ctorDef = nullRefType.GetDefaultConstructor();
+                builder.PositionAtEnd(throwBlock);
 
-                HandleDirectCall(ctorDef, ctorDef.Signature, new StackEntry[] { exceptionEntry }, null, default(LLVMValueRef), 0, NullRefFunction.GetParam(0), builder, false, default(LLVMValueRef), ctorDef);
+                ThrowException(builder, "Internal.Runtime.CompilerHelpers", "MathHelpers", "ThrowCkFiniteExc", llvmCheckFunction);
```

Comment:
I would just use the existing `ThrowHelpers.ThrowOverflowException`

Reply:
I hope I'm not sending you round in circles on this.  I think I've got it right...